core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/async/SplittingTransformer.java

```diff
@@ -219,6 +231,16 @@ private void handleCancelState() {
         }
     }
 
+    private void handleFutureCancel(Throwable e) {
```

Comment:
Pardon my ignorance, why do we have two different methods that seem like they should do the same thing? `handleCancelState` and `handleFutureCancel`

Reply:
The reason is because`handleCancelState` is for when the subscriber calls `cancel()` on the subscription. In that case we must wait until all data currently being processed by the upstream AsyncResponseTransformer is completed. `handleFutureCancel`  is for the case when the `returnFuture` is cancelled, in which case we need to stop sending data to the upstream AsyncResponseTransformer.

Reply:
Let's add some documentation for this since it's not straightforward.

Reply:
Okay that makes sense. Can `handleCancelState` be renamed to something more clear like `handleSubscriberCancel`?